sdk/samples/iot/paho_iot_pnp_with_provisioning_sample.c

```diff
@@ -0,0 +1,1218 @@
+  char property_document_topic_buffer[128];
+  rc = az_iot_pnp_client_property_document_get_publish_topic(
+      &pnp_client,
+      get_request_id(),
+      property_document_topic_buffer,
+      sizeof(property_document_topic_buffer),
+      NULL);
+  IOT_SAMPLE_EXIT_IF_AZ_FAILED(rc, "Failed to get the property document topic");
+
+  // Publish the property document request.
+  publish_mqtt_message(property_document_topic_buffer, AZ_SPAN_EMPTY, IOT_SAMPLE_MQTT_PUBLISH_QOS);
+}
+
+static void receive_messages(void)
+{
+  char* topic = NULL;
+  int topic_len = 0;
+  MQTTClient_message* message = NULL;
+  uint8_t timeout_counter = 0;
+
+  // Continue to receive commands or device property messages while device is operational.
+  while (is_device_operational)
+  {
+    IOT_SAMPLE_LOG(" "); // Formatting
+    IOT_SAMPLE_LOG("Waiting for command request or device property message.\n");
+
+    // MQTTCLIENT_SUCCESS or MQTTCLIENT_TOPICNAME_TRUNCATED if a message is received.
+    // MQTTCLIENT_SUCCESS can also indicate that the timeout expired, in which case message is NULL.
+    // MQTTCLIENT_TOPICNAME_TRUNCATED if the topic contains embedded NULL characters.
+    // An error code is returned if there was a problem trying to receive a message.
+    int rc = MQTTClient_receive(mqtt_client, &topic, &topic_len, &message, MQTT_TIMEOUT_RECEIVE_MS);
+    if ((rc != MQTTCLIENT_SUCCESS) && (rc != MQTTCLIENT_TOPICNAME_TRUNCATED))
+    {
+      IOT_SAMPLE_LOG_ERROR("Failed to receive message: MQTTClient return code %d.", rc);
+      exit(rc);
+    }
+    else if (message == NULL)
+    {
+      // Allow up to MQTT_TIMEOUT_RECEIVE_MAX_COUNT before disconnecting.
+      if (++timeout_counter >= MQTT_TIMEOUT_RECEIVE_MAX_MESSAGE_COUNT)
+      {
+        IOT_SAMPLE_LOG(
+            "Receive message timeout expiration count of %d reached.",
+            MQTT_TIMEOUT_RECEIVE_MAX_MESSAGE_COUNT);
+        return;
+      }
+    }
+    else
+    {
+      IOT_SAMPLE_LOG_SUCCESS("Client received a message from the service.");
+      timeout_counter = 0; // Reset
+
+      if (rc == MQTTCLIENT_TOPICNAME_TRUNCATED)
+      {
+        topic_len = (int)strlen(topic);
+      }
+
+      on_message_received(topic, topic_len, message);
+      IOT_SAMPLE_LOG(" "); // Formatting
+
+      MQTTClient_freeMessage(&message);
+      MQTTClient_free(topic);
+    }
+
+    send_telemetry_message();
+  }
+}
+
+static void disconnect_mqtt_client_from_iot_hub(void)
+{
+  int rc = MQTTClient_disconnect(mqtt_client, MQTT_TIMEOUT_DISCONNECT_MS);
+  if (rc != MQTTCLIENT_SUCCESS)
+  {
+    IOT_SAMPLE_LOG_ERROR("Failed to disconnect MQTT client: MQTTClient return code %d.", rc);
+    exit(rc);
+  }
+
+  MQTTClient_destroy(&mqtt_client);
+}
+
+static az_span get_request_id(void)
+{
+  az_span remainder;
+  az_span out_span = az_span_create(
+      (uint8_t*)connection_request_id_buffer, sizeof(connection_request_id_buffer));
```

Comment:
Should we store this span in a static instead of creating it here?

Reply:
ehh we might but I'd say it's not doing too much harm here.